src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16LittleEndianEncoder.cs

```diff
@@ -2,102 +2,204 @@
+
+                    uint lowSurrogate = utf16[++i];
+                    if (!IsLowSurrogate(lowSurrogate))
+                        return false;
+
+                    unchecked
+                    {
+                        codePoint -= Utf16HighSurrogateFirstCodePoint;
+                        lowSurrogate -= Utf16LowSurrogateFirstCodePoint;
+                    }
+
+                    // High surrogate contains 10 high bits of the code point
+                    codePoint = ((codePoint << 10) | lowSurrogate) + 0x010000u;
+                    charactersConsumed++;
                 }
-                codePointValue |= lowSurrogate;
-            }
 
-            codePoint = (UnicodeCodePoint)codePointValue;
+                utf32[charactersWritten++] = codePoint;
+                charactersConsumed++;
+            }
 
             return true;
         }
 
-        public static bool TryEncodeCodePoint(UnicodeCodePoint codePoint, Span<byte> buffer, out int encodedBytes)
+        #endregion Decoding implementation
+
+        #region Encoding implementation
+
+        /// <summary>
+        /// Encodes a span of UTF-32 characters into UTF-16.
+        /// 
+        /// This method will consume as many of the input characters as possible.
+        ///
+        /// On successful exit, the entire input was consumed and encoded successfully. In this case, <paramref name="charactersConsumed"/> will be
+        /// equal to the length of the <paramref name="utf32"/> and <paramref name="charactersWritten"/> will equal the total number of bytes written to
+        /// the <paramref name="utf16"/>.
+        /// 
+        /// On unsuccessful exit, the following conditions can exist.
+        ///  1) If the output buffer has been filled and no more input characters can be encoded, another call to this method with the input sliced to
+        ///     exclude the already encoded characters (using <paramref name="charactersConsumed"/>) and a new output buffer will continue the encoding.
+        ///  2) Encoding may have also stopped because the input buffer contains an invalid sequence.
+        /// </summary>
+        /// <param name="utf32">A span containing a sequence of UTF-32 characters.</param>
+        /// <param name="utf16">A span to write the UTF-16 encoded data into.</param>
+        /// <param name="charactersConsumed">On exit, contains the number of code points that were consumed from the UTF-32 character span.</param>
+        /// <param name="charactersWritten">An output parameter to store the number of bytes written to <paramref name="utf8"/></param>
+        /// <returns>True if the input buffer was fully encoded into the output buffer, otherwise false.</returns>
+        public static bool TryEncode(ReadOnlySpan<uint> utf32, Span<char> utf16, out int charactersConsumed, out int charactersWritten)
         {
-            if (!UnicodeCodePoint.IsSupportedCodePoint(codePoint))
+            charactersConsumed = 0;
```

Comment:
I noticed a performance increase if you write to a local variable and copy over to the out parameter right before returning rather than constantly writing to the out parameter directly.